### PR TITLE
Preserve the precision of a small value near zero

### PR DESCRIPTION
**Pull request by jkwak-work**
Fixes an issue #3490

When a very small value is emitted, it was snapping to a zero value. This change preserves the precision more to keep the value from being a zero value.

When "ostringstream" was printing a floating point value, we set "presicion" that indicates the number of digits in decimal. We have been using a fixed value, 20, which limits the available value near zero to be 0.00000000000000000001. When a given value is smaller, the mantissa part was truncated and it became a zero.

This change calculates how much precision is needed based on the given value and the type. The mantissa bits of "float" type can represent upto 8 digits and that of "double" type can represent upto 16 digits in decimal. The exponential part of "float" can represent upto 38 digits and that of "double" can represent upto 308 digits.

### CODE REVIEW

**Comment by csyonghe** [comment]
This will work, but emitting 100+digits for a float lit seems silly. We should switch to scientific notation when necessary.

File: source/slang/slang-emit-source-writer.cpp [hunk 1]
```diff
@@ -251,7 +252,13 @@ void SourceWriter::emit(double value)
     std::ostringstream stream;
     stream.imbue(std::locale::classic());
     stream.setf(std::ios::fixed, std::ios::floatfield);
-    stream.precision(20);
+
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/slang/slang-emit-source-writer.cpp:255]: The default behavior of ostringstream is what you described.
Because we set it to be "fixed" at the line above, the output value is always in the fixed format and never in the scientific format.
I assumed that we are using "fixed" format for a reason but I guess we can remove it judging by your comment.

I will make some changes in a way not to use "std::ios::fixed" flag.

[the remark this answers is not in the document]
Comment (csyonghe) [reply source/slang/slang-emit-source-writer.cpp:255]: The default setting won't work either because it doesn't emit enough digits when in scientific notation. We still need at least 20 digits when in scientific notation.

**Comment by jkwak-work** [comment]
I removed the use of ostringstream and implemented a function to emit float value manually.
I also updated the test case to check some edge cases.

File: source/slang/slang-emit-source-writer.cpp [hunk 2]
```diff
@@ -239,35 +239,197 @@ void SourceWriter::emit(double value)
...
-    auto slice = UnownedStringSlice(str.c_str());
-    // Remove redundant trailing 0s.
-    if (slice.end() > slice.begin())
+
+    char result[30];
+    char *next = result;
+    double valueAbs = value;
+
+    // Handle negative sign
+    const long long* d2l = reinterpret_cast<long long*>(&value);
```
Comment (csyonghe) [inline source/slang/slang-emit-source-writer.cpp:248]: Type punning like this is C++ UB, consider:
```
uint64_t uintVal = 0;
memcpy(&uintVal, &value, sizeof(uint64_t);
```

File: source/slang/slang-emit-source-writer.cpp [hunk 3]
```diff
@@ -224,7 +224,7 @@ void SourceWriter::emit(UInt64 value)
     emit(buffer);
 }
 
-void SourceWriter::emit(double value)
+inline static void SourceWriter_emit(SourceWriter *This, double value, int maxFracDigits)
```
Comment (csyonghe) [inline source/slang/slang-emit-source-writer.cpp:227]: naming convention: we don't use underscores as decl names. This can be
`static void _emitFloatingPointLiteral()`

File: source/slang/slang-emit-source-writer.cpp [hunk 4]
```diff
@@ -239,35 +239,197 @@ void SourceWriter::emit(double value)
...
-    // Remove redundant trailing 0s.
-    if (slice.end() > slice.begin())
+
+    char result[30];
+    char *next = result;
+    double valueAbs = value;
+
+    // Handle negative sign
+    const long long* d2l = reinterpret_cast<long long*>(&value);
+    const bool negative = (*d2l < 0);
```
Comment (csyonghe) [inline source/slang/slang-emit-source-writer.cpp:249]: Can this just be `sign(value)`?

File: source/slang/slang-emit-source-writer.cpp [hunk 5]
```diff
@@ -239,35 +239,197 @@ void SourceWriter::emit(double value)
...
+    else if (value == 0)
+    {
+        *next++ = '0';
+        *next++ = '.';
+        *next++ = '0';
+    }
+    else
+    {
+        int exponent = 0;
+        double mantissa = valueAbs;
```
Comment (csyonghe) [inline source/slang/slang-emit-source-writer.cpp:278]: Can we use `frexp` to extract the mantissa and exponent?

File: source/slang/slang-emit-source-writer.cpp [hunk 6]
```diff
@@ -239,35 +239,197 @@ void SourceWriter::emit(double value)
...
+        const double rounding = std::pow(10, -maxFracDigits) / 2;
+
+        // Normalize the mantissa
+        while (mantissa >= 10.0) {
```
Comment (csyonghe) [inline source/slang/slang-emit-source-writer.cpp:282]: Instead of implementing all these logic ourselves, can we call `ostream << std::scientific << std::setprecision(precision) << value` to emit a float with specific number of digits in scientific notation?

File: source/slang/slang-emit-source-writer.cpp (hunk 4 above)
Comment (jkwak-work) [reply source/slang/slang-emit-source-writer.cpp:249]: The tricky part is most of attempt to get the sign-ness will fail if the value is NaN.
I see that there is a C++11 function for this, signbit.
I will use it https://en.cppreference.com/w/cpp/numeric/math/signbit

File: source/slang/slang-emit-source-writer.cpp (hunk 6 above)
Comment (jkwak-work) [reply source/slang/slang-emit-source-writer.cpp:282]: Currently the implementation uses "fixed" format and if I switch to "scientific" format, it breaks a few existing tests.
And simple values will become harder to read. As an example, "1.0" with precision value 20 will become, "1.00000000000000000000e+00".
To void the problem, I will have to check the exponent value and decide if the value is better with scientific format or fixed format. In other words, half of my current implementation will still be there.

Note that there was a previous comment saying,
"// Eventually, the right move here would be to implement proper floating-point number formatting ourselves, "
And I agree to this statement in three reasons.
1. we will have more explicit control over how floating values are emitted.
2. our coding guide line stated not to use iostream.
3. the performance will be better because each call to the member function of ostringstream will internally take a mutex lock to prevent the system wide locale change in the middle of text parsing. I have seen from my previous project that this mutex locking cost almost 30% of the time for processing the text messages.

Comment (csyonghe) [reply source/slang/slang-emit-source-writer.cpp:282]: How about this:
1. use isnan / isinf to check for nan and infs, and early out those cases.
2. use frexp to extract exponent bits.
3. if exponent is less than a small value, use fixed output. truncate 0s in the end.
4. otherwise, use scientific output, truncate 0s before "e".

I want to avoid implementing our own logic as much as possible, because unless we also roll an extensive test suite on this, we can't guarantee its correctness.

Comment (jkwak-work) [reply source/slang/slang-emit-source-writer.cpp:282]: I will change to your suggestions.

**Comment by expipiplus1** [comment]
It's a shame that not all of our targets support hex float notation...
